server/src/main/java/org/elasticsearch/index/query/functionscore/ScoreFunctionBuilder.java

```diff
@@ -70,10 +73,17 @@ public final void writeTo(StreamOutput out) throws IOException {
      */
     @SuppressWarnings("unchecked")
     public final FB setWeight(float weight) {
-        this.weight = weight;
+        this.weight = checkWeight(weight);
         return (FB) this;
     }
 
+    private Float checkWeight(Float weight) {
+        if (weight != null && Float.compare(weight, 0) < 0) {
+            DEPRECATION_LOGGER.deprecated("[weight] cannot be negative for a filtering function");
```

Comment:
s/`filtering function`/`function score query` ? I'll fix on master ;)
Also can you put a more descriptive message about the deprecation. Something like `Setting a negative weight is deprecated and will throw an error in the next major version`.